### PR TITLE
CIRC-2122 Bypass pickup SP check for Delivery requests

### PR DESCRIPTION
**Pull request by Maksat-Galymzhan**
## Purpose
Error appears when attempting to save request with delivery to allowed Service Points

## Approach
check if fulfillment preference is Delivery
Resolves: [CIRC-2122](https://folio-org.atlassian.net/browse/CIRC-2122)

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/CreateRequestService.java [hunk 1]
```diff
@@ -231,6 +231,11 @@ private CompletableFuture<Result<RequestAndRelatedRecords>> checkPolicy(
       return completedFuture(checkPolicyForTitleLevelHold(records));
     }
 
+    if (request.isFulfillable() && request.getfulfillmentPreferenceName().equals("Delivery")){
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/domain/CreateRequestService.java:234]:
```suggestion
    if (request.isFulfillable() && DELIVERY_FULFILLMENT_PREFERENCE.equals(
    request.getfulfillmentPreferenceName()) {
```

File: src/main/java/org/folio/circulation/domain/CreateRequestService.java [hunk 2]
```diff
@@ -231,6 +231,11 @@ private CompletableFuture<Result<RequestAndRelatedRecords>> checkPolicy(
...
+      log.info("checkPolicy:: fulfillment preference is Delivery");
+      return completedFuture(succeeded(records));
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/domain/CreateRequestService.java:236]:
```suggestion
      return ofAsync(records);
```

Comment (Maksat-Galymzhan) [reply src/main/java/org/folio/circulation/domain/CreateRequestService.java]: thank for noticing it, fixed it

File: src/main/java/org/folio/circulation/domain/CreateRequestService.java [hunk 3]
```diff
@@ -231,6 +232,11 @@ private CompletableFuture<Result<RequestAndRelatedRecords>> checkPolicy(
       return completedFuture(checkPolicyForTitleLevelHold(records));
     }
 
+    if (DELIVERY == request.getfulfillmentPreference()) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/CreateRequestService.java:235]: We still need to fail if request type is not allowed by policy, so you are exiting too early. I think you need to perform this check in `org.folio.circulation.domain.RequestServiceUtility#refuseWhenRequestCannotBeFulfilled`. Maybe making sure that fulfillment preference is `Hold Shelf` before this check:
```
if (!requestPolicy.allowsServicePoint(requestType, request.getPickupServicePointId()))
```

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 4]
```diff
@@ -4979,6 +4979,51 @@ void printAndFetchRequestWithPrintEventFeatureDisabled() {
     assertThat("printDetails should be null for request2 because the print event feature is disabled",
       requestRepresentation2.getJsonObject("printDetails"), Matchers.nullValue());
   }
+
+  @Test
+  void itemLevelRequestCreationShouldFailDueToPickUpServicePointNotAllowed() {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:4984]: Is that what really happens in this test?

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 5]
```diff
@@ -4979,6 +4979,51 @@ void printAndFetchRequestWithPrintEventFeatureDisabled() {
...
+
+    Response response = requestsClient.attemptCreate(new RequestBuilder()
+      .itemRequestLevel()
+      .withFulfillmentPreference("Delivery")
+      .withRequestType(PAGE.getValue())
+      .withInstanceId(item.getInstanceId())
+      .withHoldingsRecordId(item.getHoldingsRecordId())
+      .withItemId(item.getId())
+      .withRequesterId(usersFixture.charlotte().getId())
+      .by(charlotte)
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:5017]: I think these two lines do the same thing

File: src/main/java/org/folio/circulation/domain/RequestServiceUtility.java [hunk 6]
```diff
@@ -80,6 +81,11 @@ static Result<RequestAndRelatedRecords> refuseWhenRequestCannotBeFulfilled(
       return failureDisallowedForRequestType(requestType);
     }
 
+    if (DELIVERY == request.getfulfillmentPreference()) {
+      log.info("checkPolicy:: fulfillment preference is Delivery");
+      return succeeded(requestAndRelatedRecords);
+    }
+
     if (!requestPolicy.allowsServicePoint(requestType, request.getPickupServicePointId())) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/RequestServiceUtility.java:89]:
```suggestion
    if (HOLD_SHELF == request.getfulfillmentPreference() && !requestPolicy.allowsServicePoint(
    requestType, request.getPickupServicePointId())) {
```

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 4 above)
Comment (Maksat-Galymzhan) [reply src/test/java/api/requests/RequestsAPICreationTests.java:4984]: sorry, forgot to rename it

**Review by OleksandrVidinieiev — APPROVED** [review]

File: src/main/java/org/folio/circulation/domain/RequestServiceUtility.java [hunk 7]
```diff
@@ -80,7 +81,8 @@ static Result<RequestAndRelatedRecords> refuseWhenRequestCannotBeFulfilled(
       return failureDisallowedForRequestType(requestType);
     }
 
-    if (!requestPolicy.allowsServicePoint(requestType, request.getPickupServicePointId())) {
+    if (HOLD_SHELF == request.getfulfillmentPreference() &&
+      !requestPolicy.allowsServicePoint(requestType, request.getPickupServicePointId())) {
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/domain/RequestServiceUtility.java:85]:
```suggestion
    if (HOLD_SHELF == request.getfulfillmentPreference() && !requestPolicy.allowsServicePoint(
      requestType, request.getPickupServicePointId())) {
      
```

Comment (Maksat-Galymzhan) [reply src/main/java/org/folio/circulation/domain/RequestServiceUtility.java]: fixed

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**